oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	// Convert the header and payload map to json.
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		panic("Failed to convert JWT header to json.")
+	}
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		panic("Failed to convert JWT payload to json.")
+	}
+
+	// Convert the header and payload json to base64
+	headerB64 := base64Encode(headerJson)
+	payloadB64 := base64Encode(payloadJson)
+
+	// The first two segments of the JWT are signed. The signature is the third
+	// segment.
+	segments := headerB64 + "." + payloadB64
+
+	// Sign the hash, instead of the actual segments.
+	hash := sha256.New()
+	hash.Write([]byte(segments))
+	hashed := hash.Sum(nil)
+	signerOpts := SignerOpts{}
+	signedBytes, err := signer.Sign(rand.Reader, hashed, crypto.SignerOpts(signerOpts))
+	if err != nil {
+		panic(fmt.Sprintf("Failed to sign JWT: %v", err))
+	}
+
+	// Convert the signature to base64
+	signatureB64 := base64Encode(signedBytes)
+
+	// The final JWT has three segments.
+	return segments + "." + signatureB64
+}
+
+// Run the two-legged oauth flow, will create a JWT token and use the JWT token
+// to fetch an oauth token.
+func run2LOFlow(secret map[string]interface{}, scopes []string) map[string]interface{} {
+	// Read the private key in service account secret.
+	pemBytes := []byte(secret["private_key"].(string))
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		panic("Failed to read private key.")
+	}
+
+	pkcs8key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		panic(fmt.Sprintf("Failed to decode pkcs8 private key: %v", err))
+	}
+
+	// Create a Signer object containing the private key. The Signer object has
+	// a sign function to sign a hash.
+	var signer Signer
+	if pkey, ok := pkcs8key.(*rsa.PrivateKey); ok {
+		signer = RsaSigner{pkey: pkey}
+	} else if pkey, ok := pkcs8key.(*ecdsa.PrivateKey); ok {
+		signer = EcdsaSigner{pkey: pkey}
+	} else {
+		panic("Failed to parse private key.")
+	}
+
+	// Get the JWT token
+	joinedScopes := scopesToString(scopes)
+	jwt := createJWT(secret, joinedScopes, signer)
+
+	// Construct the POST request to fetch the OAuth token.
+	postData := map[string]string{
+		"assertion":  jwt,
+		"grant_type": jwtBearerUrn,
+	}
+	params := url.Values{}
+	for key := range postData {
+		params.Add(key, postData[key])
+	}
+
+	// Send the POST request
+	response, err := http.PostForm(secret["token_uri"].(string), params)
+	if err != nil {
+		panic("Failed in http request to get service account token.")
+	}
+	defer response.Body.Close()
+
+	// The oauth token will be in the response body json.
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		panic("Failed to read http body for service account token request.")
+	}
+	var token map[string]interface{}
+	if err := json.Unmarshal(body, &token); err != nil {
+		panic("Failed to parse json.\n")
+	}
+	return token
+}
+
+// Get an OAuth2 token given a client secret json or service account json.
```

Comment:
Get an OAuth 2 access token for the specified OAuth client info and scopes.
clientInfo: a JSON text that represents either an OAuth client ID or a service account.
scopes: a list of OAuth scopes. If the scope is not a normal URL, it will be prefixed with xxx.
return: the return value will be xxx.